### PR TITLE
Xpetra: ETI for CrsMatrixWrap (try 2) 

### PR DESCRIPTION
**Pull request by jhux2**
Partially address issue #4986.

### CODE REVIEW

File: packages/xpetra/sup/CMakeLists.txt [hunk 1]
```diff
@@ -12,12 +12,63 @@ TRIBITS_CONFIGURE_FILE(${PACKAGE_NAME}_config.hpp)
...
+  SET(SOURCES ${XPETRA_ETI_CPP_SOURCES})
+  # grab any hand rolled ETI source files
+  APPEND_GLOB(SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/Utils/ExplicitInstantiation/*.cpp)
+  
+ENDIF() # if XPETRA_ENABLE_EXPLICIT_INSTANTIATION
+
 SET(HEADERS "")
 APPEND_GLOB(HEADERS */*.hpp)
-APPEND_GLOB(HEADERS */*.cpp)
+APPEND_GLOB(HEADERS */*.cpp) #JHU 25-Apr-2019 FIXME is this correct?
```
Comment (mhoemmen) [inline packages/xpetra/sup/CMakeLists.txt]: Wouldn't you want `.cpp` files to go into `SOURCES`, so they can go into the list of files to compile into the library (see `TRIBITS_ADD_LIBRARY` invocation below)?

File: packages/xpetra/sup/CMakeLists.txt [hunk 2]
```diff
@@ -12,12 +12,63 @@ TRIBITS_CONFIGURE_FILE(${PACKAGE_NAME}_config.hpp)
...
 
+# FIXME JHU is this still needed?
 SET(SOURCE mac.cpp) # Avoid empty library (MacOs 'ar' utility doesn't work on empty library).
```
Comment (mhoemmen) [inline packages/xpetra/sup/CMakeLists.txt]: It shouldn't even be right -- I don't see where `SOURCE` gets used.

Comment (jjellio) [reply packages/xpetra/sup/CMakeLists.txt]: `SET(SOURCE mac.cpp)` => `SET(SOURCES mac.cpp)`

**Comment by jjellio** [comment]
Let's say you disable ETI.   What CMake stuff takes in the 'decl' and generates Foo.hpp (with include decl; include def as its contents)?  I don't understand how Trilinos handles the ETI 'off' case.  It seems in all cases, `Foo.hpp` should always be auto generated if there is a `Foo_decl.hpp`

**Comment by mhoemmen** [comment]
@jjellio wrote:
> Let's say you disable ETI. What CMake stuff takes in the 'decl' and generates Foo.hpp (with include decl; include def as its contents)?

https://github.com/trilinos/Trilinos/blob/master/cmake/tribits/core/package_arch/TribitsCreateClientTemplateHeaders.cmake

**Comment by jhux2** [comment]
@trilinos/framework @ZUUL42 @bartlettroscoe PR testing uses "ninja -w dupbuild=err".  ninja was issuing a warning along the lines of
```
ninja: warning: multiple rules generate /scratch/jhu/trilinos/Trilinos/packages/xpetra/src/Utils/ExplicitInstantiation/Xpetra_ETI_SC_LO_GO_NO.tmpl.  builds involving this target will not be correct; continuing anyway [-w dupbuild=warn]
```
but this isn't captured in the output logs.

It seems that the PR tester should stop immediately on such a ninja error and print it to a log.  Of course, I don't know how hard implementing this might be.

Catching this would have been really difficult without @william76's jsim tool!  Thanks, @william76!

**Comment by bartlettroscoe** [comment]
> "ninja -w dupbuild=err"

Never seen that option.  Why out PR testing set that?

**Pull request closed without merging**